bridges/snowbridge/primitives/router/src/outbound/v2/mod.rs

```diff
@@ -0,0 +1,718 @@
+);
+
+impl<UniversalLocation, EthereumNetwork, OutboundQueue, AgentHashedDescription, ConvertAssetId>
+	ExportXcm
+	for EthereumBlobExporter<
+		UniversalLocation,
+		EthereumNetwork,
+		OutboundQueue,
+		AgentHashedDescription,
+		ConvertAssetId,
+	>
+where
+	UniversalLocation: Get<InteriorLocation>,
+	EthereumNetwork: Get<NetworkId>,
+	OutboundQueue: SendMessage<Balance = u128>,
+	AgentHashedDescription: ConvertLocation<H256>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+{
+	type Ticket = (Vec<u8>, XcmHash);
+
+	fn validate(
+		network: NetworkId,
+		_channel: u32,
+		universal_source: &mut Option<InteriorLocation>,
+		destination: &mut Option<InteriorLocation>,
+		message: &mut Option<Xcm<()>>,
+	) -> SendResult<Self::Ticket> {
+		let expected_network = EthereumNetwork::get();
+		let universal_location = UniversalLocation::get();
+
+		if network != expected_network {
+			log::trace!(target: TARGET, "skipped due to unmatched bridge network {network:?}.");
+			return Err(SendError::NotApplicable)
+		}
+
+		// Cloning destination to avoid modifying the value so subsequent exporters can use it.
+		let dest = destination.clone().ok_or(SendError::MissingArgument)?;
+		if dest != Here {
+			log::trace!(target: TARGET, "skipped due to unmatched remote destination {dest:?}.");
+			return Err(SendError::NotApplicable)
+		}
+
+		// Cloning universal_source to avoid modifying the value so subsequent exporters can use it.
+		let (local_net, local_sub) = universal_source.clone()
+            .ok_or_else(|| {
+                log::error!(target: TARGET, "universal source not provided.");
+                SendError::MissingArgument
+            })?
+            .split_global()
+            .map_err(|()| {
+                log::error!(target: TARGET, "could not get global consensus from universal source '{universal_source:?}'.");
+                SendError::NotApplicable
+            })?;
+
+		if Ok(local_net) != universal_location.global_consensus() {
+			log::trace!(target: TARGET, "skipped due to unmatched relay network {local_net:?}.");
+			return Err(SendError::NotApplicable)
+		}
+
+		let source_location = Location::new(1, local_sub.clone());
+
+		let agent_id = match AgentHashedDescription::convert_location(&source_location) {
+			Some(id) => id,
+			None => {
+				log::error!(target: TARGET, "unroutable due to not being able to create agent id. '{source_location:?}'");
+				return Err(SendError::NotApplicable)
+			},
+		};
+
+		let message = message.clone().ok_or_else(|| {
+			log::error!(target: TARGET, "xcm message not provided.");
+			SendError::MissingArgument
+		})?;
+
+		// Inspect AliasOrigin as V2 message
+		let mut instructions = message.clone().0;
+		let result = instructions.matcher().match_next_inst_while(
+			|_| true,
+			|inst| {
+				return match inst {
+					AliasOrigin(..) => Err(ProcessMessageError::Yield),
+					_ => Ok(ControlFlow::Continue(())),
+				}
+			},
+		);
+		ensure!(result.is_err(), SendError::NotApplicable);
```

Comment:
Yeah, seems a bit weird.  

Because we're using `match_next_inst_while` macro to match against `AliasOrigin` with error return, if not matched, then fallback to V1.

https://github.com/yrong/polkadot-sdk/pull/4#discussion_r1842521248

Reply:
But doesn't this expect the result to always be an error? For example, if result is `Ok()`, this ensure will still error?

Reply:
`NotApplicable` is a special error which will try the next item(i.e. V1 Legacy) as fallback.

Reply:
Yip, got it. But if the instruction matches `_ => Ok(ControlFlow::Continue(())),`, will `ensure!(result.is_err(), SendError::NotApplicable);` not error and the handler will quit, instead of continuing because it matched for Snowbridge v2?